src/Google.Datastore.V1Beta3/DatastoreDb.cs

```diff
@@ -171,88 +208,227 @@ private DatastoreDb(string projectId, string namespaceId, DatastoreClient client
+        }
 
         /// <summary>
         /// Allocates an ID for a single incomplete key.
         /// </summary>
+        /// <remarks>This method simply delegates to <see cref="AllocateIds(IEnumerable{Key},CallSettings)"/>.</remarks>
         /// <param name="key">The incomplete key to allocate an ID for.</param>
+        /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
         /// <returns>The complete key.</returns>
-        public Key AllocateId(Key key) => AllocateIds(new[] { key })[0];
+        public Key AllocateId(Key key, CallSettings callSettings = null) => AllocateIds(new[] { key }, callSettings)[0];
 
         /// <summary>
         /// Allocates IDs for a collection of incomplete keys.
         /// </summary>
+        /// <remarks>
+        /// <para>This method simply delegates to <see cref="AllocateIds(IEnumerable{Key},CallSettings)"/>.</para>
+        /// <para>
+        /// Call settings are not supported on this call due to restrictions with methods containing a parameter array and optional parameters.
+        /// To specify options, create a collection or array explicitly, and call <see cref="AllocateIds(IEnumerable{Key},CallSettings)"/>.
+        /// </para>
+        /// </remarks>
         /// <param name="keys">The incomplete keys. Must not be null or contain null elements.</param>
         /// <returns>A collection of complete keys with allocated IDs, in the same order as <paramref name="keys"/>.</returns>
-        public IReadOnlyList<Key> AllocateIds(params Key[] keys) => AllocateIds((IEnumerable<Key>)keys);
+        public IReadOnlyList<Key> AllocateIds(params Key[] keys) => AllocateIds(keys, null);
 
         /// <summary>
         /// Allocates IDs for a collection of incomplete keys.
         /// </summary>
         /// <param name="keys">The incomplete keys. Must not be null or contain null elements.</param>
+        /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
         /// <returns>A collection of complete keys with allocated IDs, in the same order as <paramref name="keys"/>.</returns>
-        public IReadOnlyList<Key> AllocateIds(IEnumerable<Key> keys)
+        public virtual IReadOnlyList<Key> AllocateIds(IEnumerable<Key> keys, CallSettings callSettings = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Allocates an ID for a single incomplete key asynchronously.
+        /// </summary>
+        /// <remarks>This method simply delegates to <see cref="AllocateIdsAsync(IEnumerable{Key},CallSettings)"/>.</remarks>
+        /// <param name="key">The incomplete key to allocate an ID for.</param>
+        /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
+        /// <returns>The complete key.</returns>
+        public async Task<Key> AllocateIdAsync(Key key, CallSettings callSettings = null)
```

Comment:
virtual?
All methods virtual, even delegating methods, for direct mocking of the whole API surface?